src/common.c

```diff
@@ -1686,9 +1686,15 @@ static int ofi_pollfds_waitall(struct ofi_pollfds *pfds,
 		}
 
 		if (!found && timeout > 0)
-			timeout -= (int) (ofi_gettime_ms() - start);
+		{
+			uint64_t now = ofi_gettime_ms();
+			timeout -= (int) (now - start);
+			start = now;
+			if (timeout < 0)
+				timeout = 0;
```

Comment:
This would be simpler as:
```
if (!found && timeout > 0) {
    timeout -= (int) (ofi_gettime_ms() - start);
    if (timeout < 0)
        break;
}
```
We already need to check if timeout < 0 here, at which point we know that we need to exit the while loop.  There's no need to set timeout = 0 in order to fail the while check, just exit.

Reply:
True enough about the break, but the original `timeout -= (int) (ofi_gettime_ms() - start);` isn't correct. Because timeout is being used to hold the remaining time to be waited, start needs to be reset each time in order to deduct the duration of the previous loop iteration rather than deducting the total of ALL loop iterations.

Reply:
Yeah, the timeouts aren't correct.  We should use the helper functions for this from ofi.h -- ofi_timeout_time() and ofi_adjust_timeout() -- which are designed specifically for this case.  That's what the other util code uses (e.g. see util_wait.c, util_cq.c, util_eq.c).  The general flow is:
```
endtime = ofi_timeout_time(timeout);
do {
    poll(..., timeout);

    if (ofi_adjust_timeout(endtime, &timeout))
        return -FI_ETIMEDOUT;

while (blah);
```

Reply:
The problem is it should not break if the original timeout is negative which means wait indefinitely. This should not happen on Linux since poll() won't return w/o event but as @dshinaberry said it doesn't hold on Windows.